### ISSUE TITLE
`flux push artifact` does not keep a consistent layer order

### ISSUE THREAD

**Issue by devantler** [fluxcd/flux2#4955]
### Describe the bug

When using the `flux push artifact` command the layers or not always ordered such that `application/vnd.cncf.flux.content.v1.tar+gzip` is the first layer. I suspect this is causing us issues, as we are double proxying flux packaged OCI images through two Harbors, and the Harbor to Harbor proxy seems to be unable to handle our the OCI artifacts where `application/vnd.cncf.flux.content.v1.tar+gzip` is not the first layer.

The proxy setup works well for our OCI images with layers in this order:

```json
{
  "layers": [
    {
      "digest": "<digest-goes-here>",
      "mediaType": "application/vnd.cncf.flux.content.v1.tar+gzip",
      "size": <size-of-layer>
    },
    {
      "digest": "<digest-goes-here>",
      "mediaType": "application/octet-stream",
      "size": <size-of-layer>
    }
  ]
}
```

But not for ones with this order:

```json
{
  "layers": [
    {
      "digest": "<digest-goes-here>",
      "mediaType": "application/octet-stream",
      "size": <size-of-layer>
    },
    {
      "digest": "<digest-goes-here>",
      "mediaType": "application/vnd.cncf.flux.content.v1.tar+gzip",
      "size": <size-of-layer>
    }
  ]
}
```

Furthermore we are able to pull all OCI images successfully from the first Harbor, and the issue only exists for the OCI images with the "wrong" order of layers, that have been proxied twice.

_I understand that this issue can probably be fixed over at Harbor as well, but given the simplicity in keeping the layers consistent such that `application/vnd.cncf.flux.content.v1.tar+gzip` is always the first, I wanted to raise the issue here first._

Is there a particular reason why the layers are not consistent? Is it a bug?

### Steps to reproduce

1. `flux push artifact` to ghcr.io (one with `application/vnd.cncf.flux.content.v1.tar+gzip` as the first layer, and one without (no way to control this, but it does not seem random))
2. Provision two clusters (Cluster A and Cluster B)
2. Install Flux and Harbor on each
3. Configure pull-through proxy caching on each cluster such that:
  - Cluster A proxies OCI images from ghcr.io to harbor.
  - Cluster B proxies OCI images from Harbor in Cluster A to Harbor  in Cluster B with the registry type set to harbor.
4. Pull the OCI artifacts from the harbor in Cluster B, and observe that the one with the "correct" order is pulled successfully, and the other fails to extract the first layer.

### Expected behavior

It should consistently possible to proxy flux packaged OCI artifacts through Harbors.

### Screenshots and recordings

Proxy on Cluster A:

<img width="778" alt="image" src="https://github.com/user-attachments/assets/1c9381b6-a0d6-45fd-8ce1-8117ccdf0082">

Proxy on Cluster B:

<img width="813" alt="Screenshot 2024-08-27 at 11 26 48" src="https://github.com/user-attachments/assets/744826d2-b1e7-4ddc-be4c-2a77686d3ba1">

### OS / Distro

Linux, AMD64 (QEMU Proxmox)

### Flux version

v2.3.0

### Flux check

► checking prerequisites
✔ Kubernetes 1.30.3 >=1.28.0-0
► checking version in cluster
✔ distribution: flux-v2.3.0
✔ bootstrapped: false
► checking controllers
✔ helm-controller: deployment ready
► ghcr.io/fluxcd/helm-controller:v1.0.1
✔ kustomize-controller: deployment ready
► ghcr.io/fluxcd/kustomize-controller:v1.3.0
✔ notification-controller: deployment ready
► ghcr.io/fluxcd/notification-controller:v1.3.0
✔ source-controller: deployment ready
► ghcr.io/fluxcd/source-controller:v1.3.0
► checking crds
✔ alerts.notification.toolkit.fluxcd.io/v1beta3
✔ buckets.source.toolkit.fluxcd.io/v1beta2
✔ gitrepositories.source.toolkit.fluxcd.io/v1
✔ helmcharts.source.toolkit.fluxcd.io/v1
✔ helmreleases.helm.toolkit.fluxcd.io/v2
✔ helmrepositories.source.toolkit.fluxcd.io/v1
✔ kustomizations.kustomize.toolkit.fluxcd.io/v1
✔ ocirepositories.source.toolkit.fluxcd.io/v1beta2
✔ providers.notification.toolkit.fluxcd.io/v1beta3
✔ receivers.notification.toolkit.fluxcd.io/v1
✔ all checks passed

### Container Registry provider

ghcr.io

### Additional context

flux source controller:

```txt
Warning  OCIArtifactLayerOperationFailed  13m (x10 over 39m)  source-controller  failed to extract the first layer from artifact: GET <omitted>: NOT_FOUND: http status code: 404, body: {"errors":[{"code":"NOT_FOUND","message":"http status code: 404, body: {\"errors\":[{\"code\":\"BLOB_UNKNOWN\",\"message\":\"blob unknown to registry\"}]}\n"}]}         
```

Harbor (the second proxy)

```txt
2024-08-27T08:51:59Z [DEBUG] [/controller/artifact/processor/processor.go:69]: the processor for media type application/vnd.cncf.flux.config.v1+json not found, use the default one                                                                                                                                                                    

2024-08-27T08:52:00Z [DEBUG] [/controller/proxy/controller.go:193]: Digest is not found in manifest list cache, key=<omitted>                                           

2024-08-27T08:52:00Z [DEBUG] [/pkg/allowlist/manager.go:75]: No CVE allowlist found for project 0, returning empty list.                                                    

2024-08-27T08:52:00Z [DEBUG] [/server/middleware/vulnerable/vulnerable.go:60][middleware="vulnerable" requestID="639df767-0ab1-4c89-8c8a-af913384f239"]: project proxy-harbor vulnerability prevention deactivated, skip the checking                                                                                                                  

2024-08-27T08:52:00Z [DEBUG] [/controller/artifact/processor/processor.go:69]: the processor for media type application/vnd.cncf.flux.config.v1+json not found, use the default

2024-08-27T08:52:01Z [DEBUG] [/server/middleware/repoproxy/proxy.go:320]: Failed to ensure tag {Repository:<omitted> Reference:<omitted> ProjectName:proxy-harbor Digest: Tag:<omitted> BlobMountRepository: BlobMountProjectName: BlobMountDigest:} , error the artifact is not ready yet, failed to tag it to <omitted>
```

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by stefanprodan** [comment]
The `flux push artifact` command creates a single layer, so there is nothing to order.

Example `crane manifest ghcr.io/fluxcd/flux-manifests:latest | jq`:

```json
{
  "schemaVersion": 2,
  "mediaType": "application/vnd.oci.image.manifest.v1+json",
  "config": {
    "mediaType": "application/vnd.cncf.flux.config.v1+json",
    "size": 233,
    "digest": "sha256:7283d97c2753b690ab52cf6f70047316c01608b576edcf98faa816d1fbeff0a5"
  },
  "layers": [
    {
      "mediaType": "application/vnd.cncf.flux.content.v1.tar+gzip",
      "size": 66859,
      "digest": "sha256:863bd85c869f677d0ade8c6f088ce83fb4edca307ec03c15a7006e8119ffec6b"
    }
  ],
  "annotations": {
    "org.opencontainers.image.created": "2024-05-13T13:56:21Z",
    "org.opencontainers.image.revision": "v2.3.0@sha1:896e0fa46d5107a05e953dd0a5261d78a145ec8c",
    "org.opencontainers.image.source": "git://github.com/fluxcd/flux2.git"
  }
}
```

**Comment by devantler** [comment]
Ah, the octet-stream layer must be related to hosting in ghcr.io. With crane everything seems in order, but the source controller still fails to extract the first layer for some reason.

This the crane output for the OCI image which does not work after passing through the second Harbor.

When inspecting it at ghcr.io:

```json
{
  "schemaVersion": 2,
  "mediaType": "application/vnd.oci.image.manifest.v1+json",
  "config": {
    "mediaType": "application/vnd.cncf.flux.config.v1+json",
    "size": 233,
    "digest": "sha256:bd361b2f2e53344aa66de796a6e5f3cf85a71024dc9fef8a9a824883acd7219b"
  },
  "layers": [
    {
      "mediaType": "application/vnd.cncf.flux.content.v1.tar+gzip",
      "size": 148826,
      "digest": "sha256:5f1a8009d2000dc40680bcbee8463d00770a3307e4e20ec7996d6e7e508a40fa"
    }
  ],
  "annotations": {
    "org.opencontainers.image.created": "2024-08-27T08:18:59Z",
    "org.opencontainers.image.revision": "@sha1:2ead4543a2753ac131d460750612ca41df278e53",
    "org.opencontainers.image.source": "https://github.com/energinet-digitalisering/oci-artifacts"
  }
}
```

And when inspecting it on the second Harbor:

```json
{
  "schemaVersion": 2,
  "mediaType": "application/vnd.oci.image.manifest.v1+json",
  "config": {
    "mediaType": "application/vnd.cncf.flux.config.v1+json",
    "size": 233,
    "digest": "sha256:bd361b2f2e53344aa66de796a6e5f3cf85a71024dc9fef8a9a824883acd7219b"
  },
  "layers": [
    {
      "mediaType": "application/vnd.cncf.flux.content.v1.tar+gzip",
      "size": 148826,
      "digest": "sha256:5f1a8009d2000dc40680bcbee8463d00770a3307e4e20ec7996d6e7e508a40fa"
    }
  ],
  "annotations": {
    "org.opencontainers.image.created": "2024-08-27T08:18:59Z",
    "org.opencontainers.image.revision": "@sha1:2ead4543a2753ac131d460750612ca41df278e53",
    "org.opencontainers.image.source": "https://github.com/energinet-digitalisering/oci-artifacts"
  }
}
```

So it seems nothing in the manifest changes, but the Flux source controller stil reports:

Warning  OCIArtifactLayerOperationFailed  13m (x10 over 39m)  source-controller  failed to extract the first layer from artifact: GET <omitted>: NOT_FOUND: http status code: 404, body: {"errors":[{"code":"NOT_FOUND","message":"http status code: 404, body: {\"errors\":[{\"code\":\"BLOB_UNKNOWN\",\"message\":\"blob unknown to registry\"}]}\n"}]}

I tried unpacking the OCI image with `flux artifact pull` at each location. Sadly I can't do it on the second harbor as it does not have TLS, and the `flux artifact pull` command does not support the `--insecure` flag yet. So I can't say what's going on there. 

I do believe my current issues are beyond this one. So I will be closing it. Thanks for pointing out `crane` as a tool to debug manifest. Helped a bunch 🙌🏻

**Comment by stefanprodan** [comment]
If for some reason the proxy injects random layers in the artifact, you could tell source-controller which one to use based on the media type:

```yaml
apiVersion: source.toolkit.fluxcd.io/v1beta2
kind: OCIRepository
spec:
  layerSelector:
    mediaType: "application/vnd.cncf.flux.content.v1.tar+gzip"
```

**Comment by devantler** [comment]
We did try that, but I think the issue lies with how the blobs are stored internally. I digged through the persistent volumes on the harbors to find some of the layers blobs were missing. For now we just proxy the connection to the first harbor, which works fine, and then we will revisit the issue at a later point when it is more business-critical.

I am pretty confident this was not a Flux issue :-)
